src/Symfony/Component/Validator/Mapping/Loader/AnnotationLoader.php

```diff
@@ -28,7 +28,7 @@ class AnnotationLoader implements LoaderInterface
 {
     protected $reader;
 
-    public function __construct(Reader $reader)
+    public function __construct(Reader $reader = null)
```

Comment:
any reason to make that implementation part of an AnnotationLoader that does not always have an annotation reader anymore compared to creating an AttributeLoader ?

Reply:
This is the same pattern I've used for #37474 as well. It allows a smooth transition from Doctrine Annotations to PHP Attributes for the user. I understand the `AnnotationLoader` as a loader that loads metadata from annotated PHP classes, however that annotation might look like. After all, it's an `AnnotationLoader` and not a `DoctrineAnnotationLoader`.

Most of the logic inside `AnnotationLoader` is not specific to any implementation of annotations. There's just that one new method of 30 LoC total specific to the Doctrine and PHP implementations. What I could do now is:

* Move those 30 LoC into a `DoctrineAnnotationLoaderStrategy` class and a `NativeAnnotationLoaderStrategy` and inject one or both of them into `AnnotationLoader`.
* Split `AnnotationLoader` into two classes and a trait with the shared logic.

Since we're still talking about 30 LoC (that's 15 LoC for each implementation), this looks like too much pain for very little gain.